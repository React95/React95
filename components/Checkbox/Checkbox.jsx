import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import check from '../../assets/pattern/check.svg';
import checkDisabled from '../../assets/pattern/check-disabled.svg';
import bgpattern from '../../assets/pattern/dropdown.png';

const Icon = styled.span`
  width: 12px;
  height: 12px;

  content: '';
  display: inline-block;

  position: absolute;
  left: 0;

  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;

  box-shadow: inset -1px -1px 0 0 #c3c7cb, inset 1px 1px 0 0 #000000,
    0.5px 0.5px 0 0.5px #ffffff;

  background-color: white;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: 7px 7px;
`;

const Text = styled.span`
  padding: 1px;
  user-select: none;
`;

const Field = styled.input`
  width: 12px;
  height: 12px;

  margin: 0;

  position: absolute;
  top: 0;
  left: 0;

  opacity: 0;

  &:focus ~ ${Text}, &:active ~ ${Text} {
    border: 1px dotted;
    padding: 0;
  }

  &:checked + ${Icon} {
    background-image: url('${check}');
  }

  &:checked &:disabled + ${Icon} {
    background-image:  url('${checkDisabled}'), url('${bgpattern}');
    background-size: 7px 7px, 1.9px 1.9px;
  }

  &:disabled + ${Icon} {
    background-color: #c0c0c0;
  }
`;

Field.displayName = 'Field';

const Label = styled.label`
  display: inline-block;
  height: 15px;

  line-height: 1.5;

  position: relative;

  margin: 4px 0;
  padding-left: 20px;

  ${({ disabled }) =>
    disabled &&
    `
    color: #868686;
    text-shadow: 0.5px 0.5px #d2d2d2;
  `};
`;

const Checkbox = ({
  children,
  label,
  value,
  style,
  checked,
  disabled,
  onChange,
  ...rest
}) => (
  <Label {...style} checked={checked} disabled={disabled}>
    <Field
      type="checkbox"
      value={value}
      onChange={onChange}
      checked={checked}
      disabled={disabled}
      {...rest}
    />
    <Icon />
    <Text>{children || label}</Text>
  </Label>
);

Checkbox.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  children: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

Checkbox.defaultProps = {
  value: undefined,
  label: 'Checkbox',
  children: '',
  checked: null,
  disabled: null,
  onChange: () => {},
};

export default Checkbox;
