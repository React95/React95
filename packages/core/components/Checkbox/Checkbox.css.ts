import check from './imgs/check.svg';
import checkDisabled from './imgs/check-disabled.svg';
import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';

export const icon = style({
  width: contract.space[12],
  height: contract.space[12],
  display: 'inline-block',
  position: 'absolute',
  left: '0',
  top: contract.space[2],
  borderLeftStyle: 'solid',
  borderLeftWidth: contract.space[1],
  borderLeftColor: contract.colors.borderDark,
  borderTopStyle: 'solid',
  borderTopWidth: contract.space[1],
  borderTopColor: contract.colors.borderDark,
  backgroundColor: contract.colors.inputBackground,
  backgroundRepeat: 'no-repeat',
  backgroundPosition: 'center center',
  backgroundSize: '7px 7px',
  boxShadow: contract.shadows.input,
});

export const text = style({
  padding: contract.space[1],
  userSelect: 'none',
});

export const field = style({
  width: contract.space[12],
  height: contract.space[12],
  margin: '0',
  position: 'absolute',
  top: contract.space[2],
  left: '0',
  opacity: 0,
  selectors: {},
});

export const label = style({
  display: 'inline-block',
  height: contract.space[15],
  lineHeight: 1.5,
  position: 'relative',
  marginBlock: contract.space[4],
  paddingLeft: contract.space[20],
  selectors: {
    '&:has(:disabled)': {
      color: contract.colors.borderDark,
      textShadow: `0.5px 0.5px ${contract.colors.borderLight}`,
    },
  },
});

globalStyle(`${field}:checked + ${icon}`, {
  backgroundImage: `url('${check}')`,
});

globalStyle(`${field}:focus ~ ${text}, ${field}:active ~ ${text}`, {
  borderWidth: contract.space[1],
  borderStyle: 'dotted',
  padding: '0',
});

globalStyle(`${field}:checked:disabled + ${icon}`, {
  backgroundImage: `url('${checkDisabled}')`,
  backgroundSize: '7px 7px, 1.9px 1.9px',
});

globalStyle(`${field}:disabled + ${icon}`, {
  backgroundColor: contract.colors.inputBackgroundDisabled,
});
