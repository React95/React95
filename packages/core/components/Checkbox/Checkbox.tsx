import React, { forwardRef } from 'react';

import * as styles from './Checkbox.css';

export type LabelProps = {
  disabled?: boolean;
};
export type CheckboxProps = {
  label?: string;
  children?: string;
  style?: React.CSSProperties;
} & LabelProps &
  React.HTMLProps<HTMLInputElement>;

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  ({ children, style, label, disabled = false, ...rest }, ref) => (
    <label style={style} className={styles.label}>
      <input
        className={styles.field}
        type="checkbox"
        disabled={disabled}
        {...rest}
        ref={ref}
      />
      <span className={styles.icon} />
      <span className={styles.text}>{children || label}</span>
    </label>
  ),
);
