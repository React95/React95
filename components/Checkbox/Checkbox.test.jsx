import React from 'react';
import Checkbox from './Checkbox';
import { render, fireEvent } from 'react-testing-library';

describe('<Checkbox />', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      const { container } = render(<Checkbox />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with disabled prop', () => {
      const { container } = render(<Checkbox disabled />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Checkbox label', () => {
    const label = 'label text';

    it('should display label prop as Checkbox label', () => {
      const { container } = render(<Checkbox label={label} />);

      expect(container.querySelector('label').textContent).toBe(label);
    });

    it('should display children as Checkbox label', () => {
      const { container } = render(<Checkbox>{label}</Checkbox>);

      expect(container.querySelector('label').textContent).toBe(label);
    });
  });

  describe('onChange prop', () => {
    it('should call onChange function when Checkbox is changed', () => {
      const onChangeMock = jest.fn();
      const { getByTestId } = render(
        <Checkbox onChange={onChangeMock} label="text" data-testid="checkbox" />
      );

      fireEvent.click(getByTestId('checkbox'));

      expect(onChangeMock).toHaveBeenCalled();
    });
  });
});
