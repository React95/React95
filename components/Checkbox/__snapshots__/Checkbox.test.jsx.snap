// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Snapshots should match snapshot with disabled prop 1`] = `
.c4 {
  width: 12px;
  height: 12px;
  content: '';
  display: inline-block;
  position: absolute;
  left: 0;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: 7px 7px;
}

.c6 {
  padding: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  width: 12px;
  height: 12px;
  margin: 0;
  position: absolute;
  top: 0;
  left: 0;
  opacity: 0;
}

.c2:focus ~ .c5,
.c2:active ~ .c5 {
  border: 1px dotted;
  padding: 0;
}

.c2:checked + .c3 {
  background-image: url('test-file-stub');
}

.c2:checked .c1:disabled + .c3 {
  background-image: url('test-file-stub'),url('test-file-stub');
  background-size: 7px 7px,1.9px 1.9px;
}

.c2:disabled + .c3 {
  background-color: #c0c0c0;
}

.c0 {
  display: inline-block;
  height: 15px;
  line-height: 1.5;
  position: relative;
  margin: 4px 0;
  padding-left: 20px;
  color: #868686;
  text-shadow: 0.5px 0.5px #d2d2d2;
}

<div>
  <label
    class="c0"
    disabled=""
  >
    <input
      class="c1 c2"
      disabled=""
      type="checkbox"
      value=""
    />
    <span
      class="c3 c4"
    />
    <span
      class="c5 c6"
    >
      Checkbox
    </span>
  </label>
</div>
`;

exports[`<Checkbox /> Snapshots should match snapshot without props 1`] = `
.c4 {
  width: 12px;
  height: 12px;
  content: '';
  display: inline-block;
  position: absolute;
  left: 0;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: 7px 7px;
}

.c6 {
  padding: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  width: 12px;
  height: 12px;
  margin: 0;
  position: absolute;
  top: 0;
  left: 0;
  opacity: 0;
}

.c2:focus ~ .c5,
.c2:active ~ .c5 {
  border: 1px dotted;
  padding: 0;
}

.c2:checked + .c3 {
  background-image: url('test-file-stub');
}

.c2:checked .c1:disabled + .c3 {
  background-image: url('test-file-stub'),url('test-file-stub');
  background-size: 7px 7px,1.9px 1.9px;
}

.c2:disabled + .c3 {
  background-color: #c0c0c0;
}

.c0 {
  display: inline-block;
  height: 15px;
  line-height: 1.5;
  position: relative;
  margin: 4px 0;
  padding-left: 20px;
}

<div>
  <label
    class="c0"
  >
    <input
      class="c1 c2"
      type="checkbox"
      value=""
    />
    <span
      class="c3 c4"
    />
    <span
      class="c5 c6"
    >
      Checkbox
    </span>
  </label>
</div>
`;
